### PR TITLE
8256011: Shenandoah: Don't resurrect finalizably reachable objects

### PR DESCRIPTION
**Pull request by rkennke**
In the weak-LRB we currently return referents when it is 'marked', that is when it's either reachable strongly or through a finalizable object. This means a finalizable object can be resurrected by Reference.get(), which is wrong. Only truly strongly reachable objects should be returned by Reference.get() during weak-reference-processing. 

I had to reconsider the way we call into runtime-LRBs from generated code for these reasons:
 - We need to distinguish phantom, weak and strong reference strength, and native vs in-heap access. Those are two orthogonal dimensions
 - We can have strong and phantom native referents, and strong and weak in-heap referents
 - Native referents are never compressed

Note that this depends on PR#1140.

Testing:
 - [x] hotspot_gc_shenandoah (x86_64, x64_32, aarch64)
 - [x] tier1 +UseShenandoahGC +ShenandoahVerify
 - [x] tier2 +UseShenandoahGC +ShenandoahVerify

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8256011](https://bugs.openjdk.java.net/browse/JDK-8256011): Shenandoah: Don't resurrect finalizably reachable objects

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)
 * [Zhengyu Gu](https://openjdk.java.net/census#zgu) (@zhengyu123 - **Reviewer**) ⚠️ Review applies to 37c977868e2f568a82023e25c93c6f8dee5db14b

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1109/head:pull/1109`
`$ git checkout pull/1109`

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp:105]: Have a chance to fix the typo, `non-strong` here?

**Review by shipilev — APPROVED** [review]
This makes sense, thanks.

**Review by zhengyu123 — APPROVED** [review]
Looks good.

File: src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp
Comment (shipilev) [inline src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp:863]: Typo: "strenght"

File: src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp
Comment (shipilev) [inline src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp:944]: Stick with newline style? Is `c_rarg1` on new line or not in this method?

File: src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp:1063]: "strong LRBs barriers" now... :)

File: src/hotspot/share/gc/shenandoah/c2/shenandoahSupport.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/c2/shenandoahSupport.cpp:984]: Why not separate "load_reference_barrier_strong" and  "load_reference_barrier_strong_narrow"?

File: src/hotspot/share/gc/shenandoah/c2/shenandoahSupport.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/c2/shenandoahSupport.cpp:2910]: Do we really want to strip bits from decorator here? Aren't we testing specific bits anyway downstream? I can imagine some downstream code in the future would be surprised not to see some bits that are actually set, but stripped here?

Is this for `LRBNode::hash/cmp` stability? It should be stripped in `hash/cmp` then.

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp:58]: Is this the same as `decorators & ON_STRONG_OOP_REF`?

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp:105]: A chance to fix typo "non-strorg" here.

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp:62]: Um. Shouldn't `ON_UNKNOWN_OOP_REF` default to strong? I.e. the access via Unsafe should probably resurrect the object, otherwise it corrupts the heap?

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp:108]: Maybe split out boolean locals for decorator tests? A matter of style, your call.

File: src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp:1066]: Is this `ShenandoahBarrierSet::is_strong_access`?

**Review by shipilev — CHANGES_REQUESTED** [review]
I have many questions :)

File: src/hotspot/share/gc/shenandoah/c2/shenandoahSupport.cpp
Comment (rkennke) [reply src/hotspot/share/gc/shenandoah/c2/shenandoahSupport.cpp:984]: Dunno. Could do that. Within one instance of JVM, we would never see both narrow/non-narrow versions, though.

File: src/hotspot/share/gc/shenandoah/c2/shenandoahSupport.cpp
Comment (rkennke) [reply src/hotspot/share/gc/shenandoah/c2/shenandoahSupport.cpp:2910]: Yes, this is for hash/cmp. In particular, I suspect we might miss optimizations that don't find two LRBs equal only because some non-relevant access-bit is set. Will move it to hash/cmp then.

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp
Comment (rkennke) [reply src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp:58]: It is the same, but in my testing I found that ON_STRONG_OOP_REF is not always set (it is the default when no other ON_XYZ_OOP_REF is specified).

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp
Comment (rkennke) [reply src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp:62]: No. ON_UNKNOWN_OOP_REF *might* access the referent field by reflection. We must do the correct thing in this situation. Normally, ON_UNKNOWN_OOP_REF is indeed called on normal/strong references, though, however we do the correct thing in this case too, because the referent would be marked and thus go through regular LRB. It doesn't hurt to call into weak for unknown oop refs.

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp
Comment (rkennke) [reply src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp:108]: Hmm yeah. Actually I am thinking how to do the HasDecorator tests as outermost tests, because those are the compile-time-tests. Would need to duplicate the whole test once for WEAK/UNKNOWN and once for PHANTOM I guess.

File: src/hotspot/share/gc/shenandoah/c2/shenandoahSupport.cpp
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/c2/shenandoahSupport.cpp:984]: I would be more straight-forward looking, AFAICS. Is there any code that actually matches those strings?

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp:58]: Ouch. That means we should never trust `ON_STRONG_OOP_REF`, and instead go for `SBS::is_strong_access`?

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp
Comment (shipilev) [reply src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp:62]: > No. ON_UNKNOWN_OOP_REF _might_ access the referent field by reflection. 

Ew. This just goes against my intuition about the default strongness. For example, `Unsafe_GetReference` calls with `ON_UNKNOWN_OOP_REF` -- does that mean we would treat that access as weak? I don't think we should.

File: src/hotspot/share/gc/shenandoah/c2/shenandoahSupport.cpp
Comment (rkennke) [reply src/hotspot/share/gc/shenandoah/c2/shenandoahSupport.cpp:984]: I don't think any code matches those strings. I will change them to be more specific.

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp
Comment (rkennke) [reply src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp:58]: Presumably, the Access machinery is supposed to fill it in when none of the others is set. In practice, it did not work for me. We could fix the access machinery too, but life's too short?

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp
Comment (rkennke) [reply src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp:62]: Yes that is the idea. Note, as explained above, it does not hurt. The alternative would be to generate code that figures out the strength *at runtime* and calls into the correct LRB. This would be complex, though. We already do it for SATB-barriers, but there it is necessary, while here it is harmless to call the weak-LRB for non-weak references.

**Review by shipilev — APPROVED** [review]
This looks fine to me. Let Zhengyu look through this thoroughly.

File: src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp
Comment (zhengyu123) [inline src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp:958]: Dose not seem that x86_32 made adjustment as x86_64. Dose it even build on x86_32?

**Review by zhengyu123 — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp
Comment (zhengyu123) [inline src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp:278]: This seems wrong for 32-bits. should be:

is_narrow  = LP64_ONLY(UseCompressedOops &&) NOT_LP64(false &&) !is_native;

File: src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp
Comment (zhengyu123) [inline src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp:237]: Don't need LP64_ONLY, aarch64 always 64-bits.

**Review by zhengyu123 — CHANGES_REQUESTED** [review]

**Review by zhengyu123 — APPROVED** [review]

**Review by shipilev — APPROVED** [review]
This looks fine to me. Good to go, assuming the tests pass.

**Review by shipilev — APPROVED** [review]
Okay, looks good.

**Pull request closed without merging**
